include/seqan3/range/container/aligned_allocator.hpp

```diff
@@ -84,72 +90,108 @@ class aligned_allocator
+     *       aware that users can overload any (global) ::operator new that might not adhere to the standard and might
+     *       cause std::bad_alloc or unaligned pointers.
+     *
      * \sa https://en.cppreference.com/w/cpp/memory/allocator/allocate
+     *
+     * ### Thread safety
+     *
+     * Thread-safe.
+     *
+     * ### Exception
+     *
+     * Strong exception guarantee.
      */
     [[nodiscard]]
-    pointer allocate(size_type n)
+    pointer allocate(size_type const n) const
     {
         constexpr size_type max_size = std::numeric_limits<size_type>::max() / sizeof(value_type);
         if (n > max_size)
-            throw std::bad_alloc();
-
-        // NOTE: On macOS glibc does not implement aligned_alloc, so we need to fallback to posix_memalign instead.
-#if defined(__APPLE__) && (!defined(_GLIBCXX_HAVE_ALIGNED_ALLOC) && !defined(_ISOC11_SOURCE))
-        void * p{};
-        if (int res = posix_memalign(&p, alignment, n * sizeof(value_type)); res == 0 && p != nullptr)
-            return static_cast<pointer>(p);
-#else
-        // NOTE:
-        // Allocate size bytes of uninitialized storage whose alignment is
-        // specified by alignment. The size parameter must be an integral
-        // multiple of alignment.
-        //
-        // Passing a size which is not an integral multiple of alignment or an
-        // alignment which is not valid or not supported by the implementation
-        // causes the function to fail and return a null pointer (C11, as
-        // published, specified undefined behavior in this case, this was
-        // corrected by DR 460).
-        // https://en.cppreference.com/w/cpp/memory/c/aligned_alloc
-        if (auto p = static_cast<pointer>(std::aligned_alloc(alignment, n * sizeof(value_type))))
-            return p;
-#endif
-
-        throw std::bad_alloc();
+            throw std::bad_alloc{};
+
+        size_t bytes_to_allocate = n * sizeof(value_type);
+        if constexpr (alignment <= __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+            return static_cast<pointer>(::operator new(bytes_to_allocate));
+        else // Use alignment aware allocator function.
+            return static_cast<pointer>(::operator new(bytes_to_allocate, static_cast<std::align_val_t>(alignment)));
     }
 
     /*!\brief Deallocates the storage referenced by the pointer p, which must be a pointer obtained by an earlier call
-     * to allocate().
+     * to seqan3::aligned_allocator::allocate.
+     *
+     * \param[in] p The pointer to the memory to be deallocated.
+     * \param[in] n The number of elements to be deallocated.
+     *
      * \details
      *
-     * The argument n must be equal to the first argument of the call to allocate() that originally produced p;
-     * otherwise, the behavior is undefined.
+     * The argument `n` must be equal to the first argument of the call to seqan3::aligned_allocator::allocate that
```

Comment:
This seems error prone? Can't we cache `n` and remove the parameter from this function?

Reply:
This is defined and documented according to the allocator interface requirements: https://en.cppreference.com/w/cpp/named_req/Allocator, and I am afraid we need to keep this interface.